packages/next/src/build/handle-externals.ts

```diff
@@ -62,12 +63,22 @@ export async function resolveExternal(
 ) {
   const esmExternals = !!esmExternalsConfig
   const looseEsmExternals = esmExternalsConfig === 'loose'
+  const isPackageExcludedFromBundling = optOutBundlingPackages.some((optOut) =>
+    request.startsWith(optOut)
+  )
 
   let res: string | null = null
   let isEsm: boolean = false
 
-  let preferEsmOptions =
-    esmExternals && isEsmRequested ? [true, false] : [false]
+  const preferEsmOptions =
+    esmExternals &&
+    isEsmRequested &&
+    // For package that marked as externals that should be not bundled,
+    // we don't resolve them as ESM since it could be resolved as async module,
+    // such as `import(external package)` in the bundle, valued as a `Promise`.
+    !isPackageExcludedFromBundling
+      ? [true, false]
+      : [false]
```

Comment:
Nit: if we inline `!isPackageExcludedFromBundling` directly as `!optOutBundlingPackages.some((optOut) => request.startsWith(optOut))` below, it will be faster as the `.some` is linear and doesn't have to run if `esmExternals && isEsmRequested &&` is already `false`.